### PR TITLE
Better stats for LottieGen

### PR DESCRIPTION
**Pull request by simeoncran**
I'm doing this to get ready for testing gradients.
Previously we had an awkward CSV output and a verbose option for LottieGen. Now we have a general DataTable that allows any data to be written to a TSV file.
Fixed some edge-case bugs I found in translation around gradients during testing.

### CODE REVIEW

File: LottieGen/Reporter.cs [hunk 1]
```diff
@@ -2,38 +2,67 @@
...
     // Writes a new line to the info stream.
     internal void WriteInfoNewLine()
     {
-        _infoStream.WriteLine();
+        InfoStream.WriteLine();
+    }
+
+    // Writes a row of data that can be retrieved later. Typically this is used to create CSV or TSV files
+    // containing information about the result of processing some Lottie files.
+    internal void WriteDataRow(string databaseName, IReadOnlyList<(string columnName, string value)> row)
```
Comment (simeoncran) [inline LottieGen/Reporter.cs]: >WriteDataRow [](start = 18, length = 12)

Rename to WriteDataTableRow so it's clearer how it relates to GetDataTables.

File: LottieGen/Reporter.cs [hunk 2]
```diff
@@ -2,38 +2,67 @@
...
+            }
+
+            database.AddRow(row);
+        }
+    }
+
+    // Returns the data for each dat table that was written.
+    internal IEnumerable<(string dataTableName, string[] columnNames, string[][] rows)> GetDataTables()
+    {
+        foreach (var (dataTableName, dataTable) in _dataTables.OrderBy(dt => dt.Key))
```
Comment (simeoncran) [inline LottieGen/Reporter.cs]: Wrap this in the lock so that every method is thread safe.

File: LottieGen/Reporter.cs [hunk 3]
```diff
@@ -2,38 +2,67 @@
...
     {
-        _infoStream.WriteLine();
+        InfoStream.WriteLine();
+    }
+
+    // Writes a row of data that can be retrieved later. Typically this is used to create CSV or TSV files
+    // containing information about the result of processing some Lottie files.
+    internal void WriteDataRow(string databaseName, IReadOnlyList<(string columnName, string value)> row)
+    {
+        lock (_dataTables)
```
Comment (simeoncran) [inline LottieGen/Reporter.cs]: >_dataTables [](start = 14, length = 11)

Add a separate lock. And add comments about the threading contract for the class.

File: LottieGen/Program.cs [hunk 4]
```diff
@@ -128,6 +129,43 @@ RunResult Run()
             }
         });
 #endif
+
+        if (_options.Languages.Contains(Lang.Stats))
+        {
+            // Write the stats. Stats for all the files are combined, so they are collected by the Reporter
+            // then written to files here after of the LottieFileProcessors have finished.
```
Comment (simeoncran) [inline LottieGen/Program.cs]: >after [](start = 42, length = 5)

typo: after all of

File: LottieGen/Reporter.cs [hunk 5]
```diff
@@ -2,38 +2,67 @@
...
+            {
+                database = new DataTable();
+                _dataTables.Add(databaseName, database);
+            }
+
+            database.AddRow(row);
+        }
+    }
+
+    // Returns the data for each dat table that was written.
```
Comment (eliezerpMS) [inline LottieGen/Reporter.cs]: >dat [](start = 33, length = 3)

Typo

File: LottieGen/DataTable.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+sealed class DataTable
```
Comment (simeoncran) [inline LottieGen/DataTable.cs]: Add a comment explaining what this is.

**Pull request merged**